File: ios-bookmind/Sources/SharedModels/Book.swift
import Foundation

public enum BookFormat: String, Codable, Sendable, CaseIterable {
    case epub
    case txt
}

public struct Book: Identifiable, Codable, Sendable, Hashable {
    public let id: BookID
    public var title: String
    public var author: String?
    public var format: BookFormat
    public var fileURL: URL
    public var addedAt: Date

    public init(
        id: BookID,
        title: String,
        author: String? = nil,
        format: BookFormat,
        fileURL: URL,
        addedAt: Date = Date()
    ) {
        self.id = id
        self.title = title
        self.author = author
        self.format = format
        self.fileURL = fileURL
        self.addedAt = addedAt
    }
}

public struct Chapter: Identifiable, Codable, Sendable, Hashable {
    public let id: ChapterID
    public let bookID: BookID
    public let index: Int
    public var title: String?
    public var contentRef: URL?

    public init(
        id: ChapterID,
        bookID: BookID,
        index: Int,
        title: String? = nil,
        contentRef: URL? = nil
    ) {
        self.id = id
        self.bookID = bookID
        self.index = index
        self.title = title
        self.contentRef = contentRef
    }
}

File: android-bookmind/app/src/main/java/com/bookmind/core/model/Book.kt
package com.bookmind.core.model

enum class BookFormat(val raw: String) {
    EPUB("epub"),
    FB2("fb2"),
    PDF("pdf"),
    TXT("txt");

    companion object {
        fun fromRaw(raw: String): BookFormat =
            entries.firstOrNull { it.raw == raw } ?: TXT

        /** Maps a file extension (without dot) to a format, defaulting to TXT. */
        fun fromExtension(ext: String): BookFormat = when (ext.lowercase()) {
            "epub" -> EPUB
            "fb2" -> FB2
            "pdf" -> PDF
            "txt" -> TXT
            else -> TXT
        }
    }
}

/** = iOS `Book`. */
data class Book(
    val id: BookID,
    val title: String,
    val author: String? = null,
    val format: BookFormat,
    val fileUri: String,
    val addedAt: Long = System.currentTimeMillis(),
    /** Null = not filed under any named shelf. */
    val shelfId: String? = null
)

/** A user-created named shelf books can be grouped under. */
data class Shelf(
    val id: String,
    val name: String,
    val colorRgb: Long,
    val sortOrder: Int = 0,
    val createdAt: Long = System.currentTimeMillis()
)

/** = iOS `Chapter`. */
data class Chapter(
    val id: ChapterID,
    val bookID: BookID,
    val index: Int,
    val title: String? = null,
    val contentRef: String? = null
)
